crates/sui-core/src/authority_active/gossip/mod.rs

```diff
@@ -177,97 +203,83 @@ where
-                    // for any peer.
-
-                    break },
+                    // No matter what happens we do not spend too much time on any peer.
+                    break;
+                },
 
                 items = &mut streamx.next() => {
                     match items {
-                        // Upon receiving a batch
-                        Some(Ok(BatchInfoResponseItem(UpdateItem::Batch(_signed_batch)) )) => {
-                            // Update the longer term sequence_number only after a batch that is signed
-                            self.max_seq = Some(_signed_batch.batch.next_sequence_number);
-                        },
-                        // Upon receiving a transaction digest we store it, if it is not processed already.
-                        Some(Ok(BatchInfoResponseItem(UpdateItem::Transaction((_seq, _digest))))) => {
-                            if !self.state.database.effects_exists(&_digest)? {
-                                queue.push(async move {
-                                    tokio::time::sleep(Duration::from_millis(EACH_ITEM_DELAY_MS)).await;
-                                    _digest
-                                });
+                        Some(Ok(BatchInfoResponseItem(UpdateItem::Batch(_signed_batch)) )) => {},
 
-                            }
+                        // Upon receiving a transaction digest, store it if it is not processed already.
+                        Some(Ok(BatchInfoResponseItem(UpdateItem::Transaction((seq, digest))))) => {
+                            if !self.state.database.effects_exists(&digest)? {
+                                // Download the certificate
+                                let response = self.client.handle_transaction_info_request(TransactionInfoRequest::from(digest)).await?;
+                                self.process_response(response).await?;
 
+                            }
+                            self.max_seq = Some(seq + 1);
                         },
-                        // When an error occurs we simply send back the error
+
+                        // Return any errors.
                         Some(Err( err )) => {
                             return Err(err);
                         },
+
                         // The stream has closed, re-request:
                         None => {
-
+                            tokio::time::sleep(Duration::from_millis(10)).await;
                             let req = BatchInfoRequest {
                                 start: self.max_seq,
                                 length: REQUEST_FOLLOW_NUM_DIGESTS,
                             };
-
-                            // Get a client
                             streamx = Box::pin(self.client.handle_batch_stream(req).await?);
                         },
                     }
                 },
-
-                digest = &mut queue.next() , if !queue.is_empty() => {
```

Comment:
I am affraid the queue has to go back in, otherwise we will be wasting masses of bandwidth